### PR TITLE
Add build-time support for lazy-loading

### PR DESCRIPTION
**Pull request by captainsafia**
This PR introduces a new build task to support identifying assemblies as lazy-loadable. To review this PR, start by looking at the tests in the `BuildLazyLoadTest.cs` file.

The `MarkLazyLoadDependencies` task implemented in `MarkLazyLoadDependencies` builds the dependency tree of lazy-loadable assemblies in the current project. Currently, this just reads the current project file for elements tagged with `BlazorLazyLoad` and marks their descendants as lazy-loadable. In the future, if we wanted to add more sophisticted lazy-loading handling, like specifing a dependency of a RCL as lazy-loadable but not the RCL itself, we could do it by changing the implementation here.

The changes in this PR allow you to add the `BlazorLazyLoad` metadata to a project reference, which will mark that project and its dependents as lazy-loadable.

```
<ItemGroup>
    <ProjectReference Include='..\rclwithnodependencies\RclWithNoDeps.csproj' BlazorLazyLoad='true'/>
</ItemGroup>"
```

You can also mark a single package as lazy-loadable.

```
<ItemGroup>
    <PackageReference Include='Newtonsoft.Json' Version='12.0.3' BlazorLazyLoad='true'/>
</ItemGroup>"
```

A user can also provide a list of individual assemblies that need to be lazy loaded.

```
<ItemGroup>
    <BlazorLazyLoad Include='VeryBigAssembly.dll' />
</ItemGroup>
```

**Alternatives Considered**

This was the third approach that I tried for this. I started a few different implementations but abandoned them for a few reasons:
- Derive the set of lazy-loaded components using some (gnarly) set operations on the different item groups (`_BlazorResolvedAssembly`, `_BlazorManagedRuntime`, `_LazyLoadAssemblies`)
    - Difficult to write and understand and prone to flakiness
- Integrating lazy-loading into the `ResolveBlazorRuntimeDependencies` task for Debug configurations and implementing a separate solution for scenarios with the linker
    - Requires two different solutions for linker/non-linker scenarios and introduces complexity to `ResolveBlazorRuntimeDependencies` task
- Resolve lazy loaded project and package references using the project system 

<hr/>

Addresses #22669

### CODE REVIEW

File: src/Components/WebAssembly/Build/test/BuildIntegrationTests/BuildLazyLoadTest.cs [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+        {
+            // Arrange
+            using var project = ProjectDirectory.Create("standalone", additionalProjects: new[] { "razorclasslibrary" });
+            project.Configuration = "Debug";
+
+            project.AddProjectFileContent(
+@"
+<ItemGroup>
+    <PackageReference Include='Newtonsoft.Json' Version='12.0.3' BlazorLazyLoad='true'/>
+    <PackageReference Include='Polly' Version='7.2.1' BlazorLazyLoad='true' />
```
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/Build/test/BuildIntegrationTests/BuildLazyLoadTest.cs:122]: If this works, great, but there's some possibility that the CI machines will refuse to restore arbitrary external packages. I guess we'll see :)

File: src/Components/WebAssembly/Build/testassets/rclwithpackages/rclwithpackages.csproj [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  <PropertyGroup>
+    <TargetFramework>netstandard2.1</TargetFramework>
+    <RazorLangVersion>3.0</RazorLangVersion>
+  </PropertyGroup>
+
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.AspNetCore.Components" Version="3.1.3"/>
+    <PackageReference Include="Microsoft.AspNetCore.Components.Web" Version="3.1.3" />
+    <PackageReference Include="Newtonsoft.Json" Version="12.0.3" />
```
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/Build/testassets/rclwithpackages/rclwithpackages.csproj:12]: Again I'm not sure if all our test environments will allow restoring arbitrary packages. Fine with me if it works though!

File: src/Components/WebAssembly/Build/src/Tasks/MarkLazyLoadDependencies.cs [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+        public override bool Execute()
+        {
+            // Add assemblies that have been explicitly set by the user in a
+            // `BlazorLazyLoad` item group.
+            //
+            // <ItemGroup>
+            //  <BlazorLazyLoad Include="SomeBigAssembly.dll" />
+            // </ItemGroup>
+            _lazyLoadAssemblies.UnionWith(ExplicitlazyLoad);
```
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/Build/src/Tasks/MarkLazyLoadDependencies.cs:50]: Cool, I didn't know about these APIs

File: src/Components/WebAssembly/Build/testassets/rclwithnodependencies/Class2.cs [hunk 4]
```diff
@@ -0,0 +1,6 @@
+namespace RazorClassLibrary
```
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/Build/testassets/rclwithnodependencies/Class2.cs:1]: If it's possible to remove the `.cs`/`.css`/`.js` files from the test RCLs and it still works correctly, then that would be ideal. AFAICT the `.js`/`.css` at least shouldn't be affecting the behavior here.

File: src/Components/WebAssembly/Build/src/targets/Blazor.MonoRuntime.targets [hunk 5]
```diff
@@ -143,6 +150,23 @@
...
       </_BlazorOutputWithTargetPath>
 
+
+     <!--
+      Once we processed the remaining targets, we need to add in any lazy-loaded
+      components by looking through the set of _BlazorManagedRuntimeAssembly and
+      picking out the assemblies that are:
+      - Included in the list of _LazyLoadedAssemblies
+      - Have not already been added to the Blazor output because they were resolved or required elsewhere
+      - OR are a root assembly if we are running with the mono linker
```
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/Build/src/targets/Blazor.MonoRuntime.targets:160]: I'm confused about this logic!

> Have not already been added to the Blazor output because they were resolved or required elsewhere

Wouldn't *all* dependency assemblies typically be added to the Blazor output and hence by excluded by the 2nd condition? The only way they wouldn't be added to the output is if the application doesn't have any chain of references to them.

I may be missing something, but I would have thought that if we're auto-detecting which dependencies are lazy, then the approach would be something like "mark all the non-lazy assemblies, and whatever remains are the lazy ones", like the "mark" phase in GC. Otherwise I'm unsure how dependencies `App -> RCL1 (lazy) -> Dep` and `App -> RCL2 -> Dep` would be treated. Of course, I might be misunderstanding some aspect of the design!

More broadly, I wonder how useful it is to auto-detect the dependencies anyway. Is there a plan for the dependency path info to be available at runtime somehow? I had thought people would write runtime logic like this to trigger the loading of lazy assemblies:

```cs
async Task OnNavigating(object sender, NavigatingEventArgs args)
{
    if (args.Url.StartsWith("admin"))
    {
        // Here we have to list the full transitive set of dependencies required in this part of the app
        await AssemblyLoader.LoadAssembliesAsync(
            "MyCompany.MyApp.Admin",
            "ComponentsVendor.BlazorComponents",
            "Newtonsoft.Json",
            "System.Memory",
            "System.Data");
    }
    else if (args.Url == "Products/Export")
    {
        // The PDF exporter is big so we only load it on this one page
        await AssemblyLoader.LoadAssembliesAsync("ComponentsVendor.PdfExport");
    }
}
```

If this is the case, then the developer still has to hard-code knowledge of the full transitive set of dependencies in their runtime code, so I'm not certain how helpful it is to auto-detect aspects of it at build time.

We do have the option of doing something higher-level where the developer doesn't specify individual assemblies to be loaded at runtime, but instead specifies something like which top-level lazy-loaded thing is to be loaded, and the system uses knowledge of the dependency graph to get all the right things loaded. However that takes us pretty close back to the design we were trying to avoid because it's more sophisticated than the min-bar and is something that developers could build on their own if they needed.

I'd be interested to know what you think is preferable here. We *can* still do the more high-level solution if you think it's going to be pretty cheap to implement, or we could consider that a possible future enhancement with the initial version simply requiring the exact full set of assemblies to be specified both at build-time and runtime.

File: src/Components/WebAssembly/Build/src/Tasks/MarkLazyLoadDependencies.cs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+            _visitedAssemblies.Add(fileName);
+
+            // If the source is a project reference and the assembly is marked as
+            // lazy then we are processing a lazily-loaded project.
+            //
+            // <ProjectReference Include="SomeProj.csproj" BlazorLazyLoad="true" />
+            var sourceFromProjectReference = assembly.GetMetadata(REFERENCE_SOURCE_KEY) == PROJECT_REFERENCE;
+            if (sourceFromProjectReference && assembly.GetMetadata(LAZY_LOAD_LABEL) == "true")
+            {
+                Log.LogMessage($"{fileName} is a lazy-loaded project reference.");
```
Comment (pranavkm) [inline src/Components/WebAssembly/Build/src/Tasks/MarkLazyLoadDependencies.cs:90]: might want to log these with low improtance (I think it defaults to normal)

File: src/Components/WebAssembly/Build/src/Tasks/MarkLazyLoadDependencies.cs [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+                _lazyLoadAssemblies.Add(new TaskItem(String.Concat(fileName, ".dll")));
```
Comment (pranavkm) [inline src/Components/WebAssembly/Build/src/Tasks/MarkLazyLoadDependencies.cs:91]:
```suggestion
                _lazyLoadAssemblies.Add(new TaskItem(fileName + ".dll"));
```

File: src/Components/WebAssembly/Build/src/Tasks/MarkLazyLoadDependencies.cs [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+            return !Log.HasLoggedErrors;
+        }
+
+        private void ProcessAssembly(ITaskItem assembly)
+        {
+            var fileName = assembly.GetMetadata("FileName");
+
+            var visited = _visitedAssemblies.Contains(fileName);
+            if (visited)
```
Comment (pranavkm) [inline src/Components/WebAssembly/Build/src/Tasks/MarkLazyLoadDependencies.cs:74]: You could replace `! .Contains` + `.Add` with

```suggestion
            if (!_visitedAssemblies.Add(fileName))
```

File: src/Components/WebAssembly/Build/test/BuildIntegrationTests/BuildLazyLoadTest.cs [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+</ItemGroup>");
+
+            var result = await MSBuildProcessManager.DotnetMSBuild(project, args: "/restore");
```
Comment (pranavkm) [inline src/Components/WebAssembly/Build/test/BuildIntegrationTests/BuildLazyLoadTest.cs:125]: We try and avoid doing restores as part of these tests. From our past experience, running restore causes test hangs \ timeouts, and we invariably end up skipping them.

File: src/Components/WebAssembly/Build/test/BuildIntegrationTests/BuildLazyLoadTest.cs [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+            Assert.FileExists(result, buildOutputDirectory, "wwwroot", "_framework", "blazor.boot.json");
+            // And that the assembly is in the output
+            Assert.FileExists(result, buildOutputDirectory, "wwwroot", "_framework", "_bin", "RclWithNoDeps.dll");
+
+            var bootJson = ReadBootJsonData(result, Path.Combine(buildOutputDirectory, "wwwroot", "_framework", "blazor.boot.json"));
+
+            // And that it has been labelled as a dynamic assembly in the boot.json
+            var dynamicAssemblies = bootJson.resources.dynamicAssembly;
+            Assert.NotNull(dynamicAssemblies);
+            Assert.Contains("RclWithNoDeps.dll", dynamicAssemblies.Keys);
```
Comment (pranavkm) [inline src/Components/WebAssembly/Build/test/BuildIntegrationTests/BuildLazyLoadTest.cs:47]: We should verify `RclWithNoDeps.dll` (and the subset of assemblies that only appear as it's dependencies) do not appear in `bootJson.resources.assembly`.

**Review by pranavkm — COMMENTED** [review]
I'm working on moving some of these targets to the RazorSDK. Looking at these changes, I think this should be easy to move it over without too many issues, but wanted to give you a head's up.

File: src/Components/WebAssembly/Build/test/BuildIntegrationTests/BuildLazyLoadTest.cs (hunk 1 above)
Comment (captainsafia) [reply src/Components/WebAssembly/Build/test/BuildIntegrationTests/BuildLazyLoadTest.cs]: `Newtonsoft.Json` should be OK to use in tests since it is included in our Dependencies files.

I think `Polly` will work either way but I've removed it, just in case.

File: src/Components/WebAssembly/Build/testassets/rclwithnodependencies/Class2.cs (hunk 4 above)
Comment (captainsafia) [reply src/Components/WebAssembly/Build/testassets/rclwithnodependencies/Class2.cs]: Done.

File: src/Components/WebAssembly/Build/src/Tasks/MarkLazyLoadDependencies.cs (hunk 6 above)
Comment (captainsafia) [reply src/Components/WebAssembly/Build/src/Tasks/MarkLazyLoadDependencies.cs]: Ended up removing the logs all together. They were helpful during development but not needed.

File: src/Components/WebAssembly/Build/src/Tasks/MarkLazyLoadDependencies.cs (hunk 8 above)
Comment (captainsafia) [reply src/Components/WebAssembly/Build/src/Tasks/MarkLazyLoadDependencies.cs]: Ended up changing the type of `_visitedAssemblies` to `HashSet` to use this API.

File: src/Components/WebAssembly/Build/test/BuildIntegrationTests/BuildLazyLoadTest.cs (hunk 10 above)
Comment (captainsafia) [reply src/Components/WebAssembly/Build/test/BuildIntegrationTests/BuildLazyLoadTest.cs:52]: Added this.

**Comment by captainsafia** [comment]
There And Back Again: A Build Newb's Tale

Thanks for the review on this one folks! Admittedly, this is a little bit more complicated than the design that we settled on during our discussions.

Implementing the standard `<BlazorLazyLoad Include="Blah.dll" />` was simple enough so I extended the exercise a little bit and tried to implement something that addressed the stretch goals we had.

I thought I had gotten something working but I've learned some things since that have changed my perception of the problem:

- Project system evaluation in tasks is not recommended which required a significant rewrite of the original design
- The test cases (and my own thinking) didn't account for the fact that the lazy-loaded dependencies will already be resolved since they will be referenced by user code
  - @SteveSandersonMS Your point about being confused by the use of the `_BlazorResolvedAssembly` reference in the condition was due to this gap
- Lots of time spent mucking around with MSBuild conditions that weren't structured properly
- The `BlazorLazyLoad` metadata attribute isn't readily available in the resolved assemblies, you have to be able to trace it back to a package reference or project reference in some way

After all this experimentation, this PR is pretty much back to supporting the basic scenarios.

> I'd be interested to know what you think is preferable here. We can still do the more high-level solution if you think it's going to be pretty cheap to implement, or we could consider that a possible future enhancement with the initial version simply requiring the exact full set of assemblies to be specified both at build-time and runtime.

I'm very interested in doing something that would avoid users having to specify all the assemblies to dynamically load at runtime. I suspect that this is the one of the big friction points that users will run into with this design.

Having done some of the work over the past few days to get closer to support that, I've got a better understanding of the complexities of this (outside of the points that were raised during the design meeting). Once I let some of the things I've learned in this first round marinate, I might be able to come up with something better for round 2.

With that in mind, I think we'll move forward with shipping the bare bones version for preview7. For .NET 5, I'd like to move the min-bar to support auto-resolving dependencies.

Tl;DR: I tried to implement above the min-bar, discovered some of the difficulties in the process, settled for min-bar, but hope to implement something a bit better for .NET 5.

File: src/Components/WebAssembly/Build/src/Tasks/MarkLazyLoadDependencies.cs [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+using System.Runtime.Serialization;
+using System.Runtime.Serialization.Json;
+using System.Text;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using ResourceHashesByNameDictionary = System.Collections.Generic.Dictionary<string, string>;
+
+namespace Microsoft.AspNetCore.Components.WebAssembly.Build
+{
+    public class MarkLazyLoadDependencies : Task
```
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/Build/src/Tasks/MarkLazyLoadDependencies.cs:20]: Should this be renamed to something like `ResolveEagerlyLoadedAssemblies` now?
